File: frontend/desktop/src/types/events.ts
import type { PipelineStage, StageStatus, JudgeVerdict } from "./pipeline";

/** Shared Tauri event names for pipeline runtime updates. */
export const PIPELINE_EVENTS = {
  started: "pipeline:started",
  stage: "pipeline:stage",
  log: "pipeline:log",
  artifact: "pipeline:artifact",
  question: "pipeline:question",
  completed: "pipeline:completed",
  error: "pipeline:error",
} as const;

// ---- Backend event payloads ----

/** Emitted when a new pipeline run begins. */
export interface PipelineStartedEvent {
  runId: string;
  sessionId: string;
  prompt: string;
  workspacePath: string;
  maxIterations: number;
}

/** Emitted when a pipeline stage changes status. */
export interface PipelineStageEvent {
  runId: string;
  stage: PipelineStage;
  status: StageStatus;
  iteration: number;
  durationMs?: number;
}

/** Emitted for each log line produced by an agent. */
export interface PipelineLogEvent {
  runId: string;
  stage: PipelineStage;
  line: string;
  stream: "stdout" | "stderr";
}

/** Emitted when a pipeline artefact is produced. */
export interface PipelineArtifactEvent {
  runId: string;
  kind: string;
  content: string;
  iteration: number;
}

/** Emitted when the pipeline completes successfully. */
export interface PipelineCompletedEvent {
  runId: string;
  verdict: JudgeVerdict;
  totalIterations: number;
  durationMs: number;
}

/** Emitted when the pipeline encounters an error. */
export interface PipelineErrorEvent {
  runId: string;
  stage?: PipelineStage;
  message: string;
}

/** Pipeline question event — emitted when the pipeline pauses for user input. */
export interface PipelineQuestionEvent {
  runId: string;
  questionId: string;
  stage: PipelineStage;
  iteration: number;
  questionText: string;
  agentOutput: string;
  optional: boolean;
}

/** Answer payload sent back to the backend via the answer_pipeline_question command. */
export interface PipelineAnswer {
  runId: string;
  questionId: string;
  answer: string;
  skipped: boolean;
}

// ---- JSONL Event Log Types ----

/** Status values for stage completion in JSONL events. */
export type StageEndStatus = "completed" | "failed" | "skipped";

/** Run status values for JSONL events and run summaries. */
export type RunStatus =
  | "running"
  | "paused"
  | "waiting_for_input"
  | "completed"
  | "failed"
  | "cancelled"
  | "crashed";

/** Union type for all events stored in events.jsonl.
 *
 * Every event has schema version (v), sequence number (seq), and
 * timestamp (ts) in RFC 3339 format. The discriminated union uses
 * the `type` field to distinguish variants.
 */
export type RunEvent =
  | RunStartEvent
  | StageStartEvent
  | StageEndEvent
  | IterationEndEvent
  | QuestionEvent
  | RunEndEvent;

/** Base fields present in all run events. */
export interface RunEventBase {
  /** Schema version for migration handling. */
  v: number;
  /** Monotonic sequence number (1-based). */
  seq: number;
  /** Timestamp in RFC 3339 format. */
  ts: string;
}

/** Marks the beginning of a pipeline run. */
export interface RunStartEvent extends RunEventBase {
  type: "run_start";
  /** The original user prompt. */
  prompt: string;
  /** Maximum iterations allowed for this run. */
  maxIterations: number;
}

/** A stage begins execution. */
export interface StageStartEvent extends RunEventBase {
  type: "stage_start";
  /** The stage being started. */
  stage: PipelineStage;
  /** Current iteration number (1-based). */
  iteration: number;
}

/** A stage finishes execution. */
export interface StageEndEvent extends RunEventBase {
  type: "stage_end";
  /** The stage that finished. */
  stage: PipelineStage;
  /** Current iteration number (1-based). */
  iteration: number;
  /** Completion status of the stage. */
  status: StageEndStatus;
  /** Duration in milliseconds. */
  durationMs: number;
  /** Judge verdict (only for judge stage). */
  verdict?: JudgeVerdict;
  /** Plan audit verdict (only for plan_audit stage). */
  auditVerdict?: "APPROVED" | "REJECTED" | "NEEDS_REVISION";
}

/** An iteration loop completes with judge verdict. */
export interface IterationEndEvent extends RunEventBase {
  type: "iteration_end";
  /** The iteration that just completed (1-based). */
  iteration: number;
  /** Judge verdict for this iteration. */
  verdict: JudgeVerdict;
}

/** User answered a question during the run. */
export interface QuestionEvent extends RunEventBase {
  type: "question";
  /** The stage that asked the question. */
  stage: PipelineStage;
  /** Current iteration when question was asked (1-based). */
  iteration: number;
  /** The question text. */
  question: string;
  /** The user's answer (may be empty if skipped). */
  answer: string;
  /** Whether the question was skipped. */
  skipped: boolean;
}

/** Terminal event — run completed, failed, or cancelled.
 *
 * Every run MUST end with exactly one RunEndEvent.
 * If no terminal event exists, the run is considered crashed.
 */
export interface RunEndEvent extends RunEventBase {
  type: "run_end";
  /** Final run status. */
  status: RunStatus;
  /** Final judge verdict (null if failed/cancelled before judge). */
  verdict: JudgeVerdict | null;
  /** Error message if status is failed. */
  error?: string;
  /** Timestamp when the crash was recovered (if recovered). */
  recoveredAt?: string;
}

File: frontend/desktop/src-tauri/src/models/events.rs
use serde::{Deserialize, Serialize};

use super::pipeline::{JudgeVerdict, PipelineStage};

/// Run status for terminal events.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RunStatus {
    Completed,
    Failed,
    Cancelled,
    Crashed,
}

/// Stage end status variants.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum StageEndStatus {
    Completed,
    Failed,
    Skipped,
}

/// Plan audit verdict variants.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum PlanAuditVerdict {
    #[serde(rename = "APPROVED")]
    Approved,
    #[serde(rename = "REJECTED")]
    Rejected,
    #[serde(rename = "NEEDS_REVISION")]
    NeedsRevision,
}

/// Individual run event for events.jsonl.
///
/// Every event includes:
/// - `v`: schema version (u32)
/// - `seq`: monotonic sequence number (u64)
/// - `ts`: RFC 3339 timestamp
///
/// The `type` field discriminates the variant.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum RunEvent {
    /// Marks the beginning of a pipeline run.
    #[serde(rename_all = "camelCase")]
    RunStart {
        v: u32,
        seq: u64,
        ts: String,
        prompt: String,
        max_iterations: u32,
    },

    /// A stage begins execution.
    #[serde(rename_all = "camelCase")]
    StageStart {
        v: u32,
        seq: u64,
        ts: String,
        stage: PipelineStage,
        iteration: u32,
    },

    /// A stage finishes execution.
    #[serde(rename_all = "camelCase")]
    StageEnd {
        v: u32,
        seq: u64,
        ts: String,
        stage: PipelineStage,
        iteration: u32,
        status: StageEndStatus,
        /// Duration in milliseconds.
        duration_ms: u64,
        /// Plan audit verdict, if this was a plan_audit stage.
        #[serde(skip_serializing_if = "Option::is_none")]
        audit_verdict: Option<PlanAuditVerdict>,
        /// Judge verdict, if this was a judge stage.
        #[serde(skip_serializing_if = "Option::is_none")]
        verdict: Option<JudgeVerdict>,
    },

    /// An iteration loop completes with judge verdict.
    #[serde(rename_all = "camelCase")]
    IterationEnd {
        v: u32,
        seq: u64,
        ts: String,
        iteration: u32,
        verdict: JudgeVerdict,
    },

    /// User answered a question during the run.
    #[serde(rename_all = "camelCase")]
    Question {
        v: u32,
        seq: u64,
        ts: String,
        stage: PipelineStage,
        iteration: u32,
        question: String,
        answer: String,
        skipped: bool,
    },

    /// Terminal event - run completed, failed, or cancelled.
    #[serde(rename_all = "camelCase")]
    RunEnd {
        v: u32,
        seq: u64,
        ts: String,
        status: RunStatus,
        #[serde(skip_serializing_if = "Option::is_none")]
        verdict: Option<JudgeVerdict>,
        /// Error message when status is failed or crashed.
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<String>,
        /// Recovery timestamp for crash recovery events.
        #[serde(skip_serializing_if = "Option::is_none")]
        recovered_at: Option<String>,
    },
}

impl RunEvent {
    /// Returns the schema version of this event.
    pub fn schema_version(&self) -> u32 {
        match self {
            RunEvent::RunStart { v, .. } => *v,
            RunEvent::StageStart { v, .. } => *v,
            RunEvent::StageEnd { v, .. } => *v,
            RunEvent::IterationEnd { v, .. } => *v,
            RunEvent::Question { v, .. } => *v,
            RunEvent::RunEnd { v, .. } => *v,
        }
    }

    /// Returns the sequence number of this event.
    pub fn sequence(&self) -> u64 {
        match self {
            RunEvent::RunStart { seq, .. } => *seq,
            RunEvent::StageStart { seq, .. } => *seq,
            RunEvent::StageEnd { seq, .. } => *seq,
            RunEvent::IterationEnd { seq, .. } => *seq,
            RunEvent::Question { seq, .. } => *seq,
            RunEvent::RunEnd { seq, .. } => *seq,
        }
    }

    /// Returns the timestamp of this event.
    pub fn timestamp(&self) -> &str {
        match self {
            RunEvent::RunStart { ts, .. } => ts,
            RunEvent::StageStart { ts, .. } => ts,
            RunEvent::StageEnd { ts, .. } => ts,
            RunEvent::IterationEnd { ts, .. } => ts,
            RunEvent::Question { ts, .. } => ts,
            RunEvent::RunEnd { ts, .. } => ts,
        }
    }

    /// Returns true if this is a terminal event (run_end).
    pub fn is_terminal(&self) -> bool {
        matches!(self, RunEvent::RunEnd { .. })
    }
}
